app/models/category.rb

```diff
@@ -0,0 +1,22 @@
+class Category < ApplicationRecord
+
+  belongs_to :parent_category, class_name: "Category", optional: true
+  has_many :products, dependent: :restrict_with_error
+  has_many :subcategories_products, through: :sub_categories, source: :products, dependent: :restrict_with_error
+  has_many :sub_categories, class_name: "Category", foreign_key: "parent_category_id", dependent: :destroy
```

Comment:
move this above `subcategories_products`

Reply:
done